executor/aggfuncs/func_count.go

```diff
@@ -0,0 +1,99 @@
+package aggfuncs
+
+import (
+	"github.com/juju/errors"
+
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type baseCount struct {
+	baseAggFunc
+}
+
+type partialResult4Count = int64
+
+func (e *baseCount) AllocPartialResult() PartialResult {
+	return PartialResult(new(partialResult4Count))
+}
+
+func (e *baseCount) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4Count)(pr)
+	*p = 0
+}
+
+func (e *baseCount) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4Count)(pr)
+	chk.AppendInt64(e.ordinal, *p)
+	return nil
+}
+
+type countOriginal struct {
+	baseCount
+}
+
+func (e *countOriginal) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4Count)(pr)
+
+	for _, row := range rowsInGroup {
+		_, isNull, err := e.args[0].EvalInt(sctx, row)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if isNull {
+			continue
+		}
+
+		*p++
+	}
+
+	return nil
+}
+
+type countOriginalWithDistinct struct {
+	baseCount
+
+	dc *distinctChecker
+}
+
+func (e *countOriginalWithDistinct) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4Count)(pr)
+
+	for _, row := range rowsInGroup {
+		buf, hasNull, err := encodeRow(sctx, e.args, row)
```

Comment:
encodeRow is expensive, [this implementation](https://github.com/pingcap/tidb/pull/7015/files#diff-b254a90c3c5e1fdbac7a00d8474132a6R126) can be referred.

Reply:
As discussed at https://github.com/pingcap/tidb/pull/7009#discussion_r200811799 , `count` may have many parameters, like `select count(distinct a, b) from t`, we need to eval the whole row to judge whether current row should be counted.

Reply:
1. Such as `select count(distinct a, b), c from t;`
The `row` will be [a, b, c]. So encode the whole row is still not correct.
2. There is already a distinctChecker, we can implement a `checkChunkRow` function for it.
3. You can implement a function like `EncodeValue4ChunkRow` in codec.go, and invoke `encodeChunkRow` in it.

Reply:
1. `encodeRow` will get the value from `row` by `e.args`.
2. I think the `checkChunkRow` is what `encodeRow` did here.
3. The problem of `encodeChunkRow` is I can't get the corrent colIdx by expressions. And it looks like I can't import `expression` in `chunk` package.

Reply:
1. The `allTypes` is the fieldTypes of count.Args.
The `colIdx` is the ordinal of count.Args, e.g. for count(a, b, c), the colIdx is [0,1,2].
2. I think we'd better change `encodeRow` to `evalOneRow` which only eval the result of the expressions and invoke EncodeValue4ChunkRow outside. If you take a look at `EncodeValue`, you can see that there are 2 parameters `comparable` and `hash`, but `encodeRow` does not consider them.

Reply:
1. I used to use this logic (invoke `HashChunkRow` in codec.go instead if `encodeRow`) but failed test `select count(distinct c), count(distinct a, b) from t`. For the `distinct c`, the ordinal is `0`, but the correct ordinal is `2`.
2. I will have a try.